src/bson_mapper/test/archiver_test.cpp

```diff
@@ -0,0 +1,514 @@
+        oarchive(test_obj);
+    }
+
+    {
+        DataE test_obj;
+        std::ifstream is("embedded_view_data_test.bson");
+        bson_mapper::BSONInputArchive iarchive(is);
+        iarchive(test_obj);
+
+        REQUIRE(countKeys(test_obj.d.getUnderlyingBSONData()) == 4);
+    }
+}
+
+TEST_CASE(
+    "the BSON archiver successfully serializes and deserializes a variety of C++ classes "
+    "containing BSON objects, std::vectors, and primitive types compatible with BSON.") {
+    {
+        std::ofstream os("data.bson", std::ios_base::binary);
+        bson_mapper::BSONOutputArchive archive(os);
+
+        DataA a1;
+        DataA a2;
+        DataB b1;
+
+        std::vector<int> test_arr;
+
+        test_arr.push_back(1);
+        test_arr.push_back(2);
+        test_arr.push_back(3);
+
+        std::vector<DataA> test_obj_arr;
+
+        a1.x = 43;
+        a1.y = 229;
+        a1.z = 3.14;
+
+        a2.x = 26;
+        a2.y = 32;
+        a2.z = 3.4;
+
+        test_obj_arr.push_back(a1);
+        test_obj_arr.push_back(a2);
+
+        b1.a = 517259871609285984;
+        b1.b = 35781926586124;
+        b1.m = a2;
+        b1.arr = test_obj_arr;
+        b1.s = "hello world!";
+        b1.tp = std::chrono::system_clock::now();
+
+        DataC c1;
+
+        int i = 10;
+
+        std::chrono::system_clock::time_point tp = std::chrono::system_clock::now();
+
+        archive(CEREAL_NVP(tp), CEREAL_NVP(test_obj_arr), CEREAL_NVP(i), a1, a2, b1, c1);
+    }
+
+    {
+        DataA a1;
+        DataA a2;
+        DataB b1;
+
+        std::vector<DataA> test_obj_arr;
+
+        DataC c1;
+        c1.resetValues();
+
+        int i;
+
+        std::chrono::system_clock::time_point tp;
+
+        std::ifstream is("data.bson");
+        bson_mapper::BSONInputArchive archive(is);
+        archive(CEREAL_NVP(tp), CEREAL_NVP(test_obj_arr), CEREAL_NVP(i), a1, a2, b1, c1);
+
+        REQUIRE(countKeys(c1.getUnderlyingBSONData()) == 21);
+
+        std::ofstream os("data_clone.bson", std::ios_base::binary);
+        bson_mapper::BSONOutputArchive oarchive(os);
+        oarchive(CEREAL_NVP(tp), CEREAL_NVP(test_obj_arr), CEREAL_NVP(i), a1, a2, b1, c1);
+    }
+
+    REQUIRE(file_cmp("data.bson", "data_clone.bson"));
+}
+
+TEST_CASE("the BSON archiver successfully serializes and deserializes a single C++ class.") {
+    DataA a1;
+    a1.x = 229;
+    a1.y = 43;
+    a1.z = 3.14159;
+
+    std::ofstream os("single_test.bson");
+    bson_mapper::BSONOutputArchive oarchive(os);
+    oarchive(a1);
+    os.close();
+
+    DataA a2;
+    std::ifstream is("single_test.bson");
+    bson_mapper::BSONInputArchive iarchive(is);
+    iarchive(a2);
+
+    REQUIRE(a1 == a2);
+}
+
+TEST_CASE(
+    "the BSON archiver enforces the existence of name-value pairs in input archives without "
+    "segfaulting.") {
+    BadDataA a1;
+    DataA a2;
+    DataB b1;
+
+    std::vector<DataA> test_obj_arr;
+
+    DataC c1;
+    c1.resetValues();
```

Comment:
Archiving this after reseting could be dangerous, see this [CXX ticket](https://jira.mongodb.org/browse/CXX-940).

If in the reset for oid's, you switched to the `bsoncxx::oid{bsoncxx::oid::init_tag}` ctor, this problem should go away?